plugins/ROS/src/plugin/ROS.cpp

```diff
@@ -33,64 +33,47 @@ void ROSPlugin::init(mc_control::MCGlobalController & controller, const mc_rtc::
 
 void ROSPlugin::reset(mc_control::MCGlobalController & controller)
 {
-  if(publish_control)
-  {
-    mc_rtc::ROSBridge::init_robot_publisher("control", controller.timestep(), controller.controller().outputRobot());
-  }
-  if(publish_env)
+
+  auto publish_env = [&controller](const std::string & prefix, mc_rbdyn::Robots & robots, bool use_real)
   {
-    auto publish_env = [&controller](const std::string & prefix, mc_rbdyn::Robots & robots, bool use_real)
+    for(const auto & r : robots)
     {
-      for(size_t i = 1; i < robots.size(); ++i)
-      {
-        mc_rtc::ROSBridge::init_robot_publisher(prefix + "_" + std::to_string(i), controller.timestep(),
-                                                robots.robot(i), use_real);
-      }
-    };
-    publish_env("control/env", controller.controller().outputRobots(), false);
-    if(publish_real) { publish_env("real/env", controller.controller().outputRealRobots(), true); }
-  }
-  if(publish_real)
-  {
-    const auto & real_robot = controller.controller().outputRealRobot();
-    mc_rtc::ROSBridge::init_robot_publisher("real", controller.timestep(), real_robot, true);
-  }
+      mc_rtc::ROSBridge::init_robot_publisher(prefix + r.name(), controller.timestep(), r, use_real);
+    }
+  };
+
+  publish_env("control/", controller.robots(), false);
+
+  if(publish_real) { publish_env("real/", controller.robots(), true); }
 }
 
 void ROSPlugin::after(mc_control::MCGlobalController & controller)
 {
-  if(publish_control)
-  {
-    mc_rtc::ROSBridge::update_robot_publisher("control", controller.timestep(), controller.controller().outputRobot());
-  }
-  // Publish environment state
-  if(publish_env)
+  // mc_rtc::ROSBridge::update_robot_publisher("control/" + controller.robot().name(), controller.timestep(),
+  // controller.controller().outputRobot());
+
+  auto update_env = [this, &controller](const std::string & prefix, mc_rbdyn::Robots & robots)
   {
-    auto update_env = [this, &controller](const std::string & prefix, mc_rbdyn::Robots & robots)
+    for(const auto & r : robots)
     {
-      for(size_t i = 1; i < robots.size(); ++i)
-      {
-        mc_rtc::ROSBridge::update_robot_publisher(prefix + "_" + std::to_string(i), controller.timestep(),
-                                                  robots.robot(i));
-      }
-      published_env = std::max<size_t>(publish_env, robots.size() - 1);
-    };
-    update_env("control/env", controller.controller().outputRobots());
-    if(publish_real) { update_env("real/env", controller.controller().outputRealRobots()); }
-  }
-  // Publish real robot
-  if(publish_real)
+      mc_rtc::ROSBridge::update_robot_publisher(prefix + r.name(), controller.timestep(), r);
+    }
+    published_topics = robots.size() - 1;
+  };
+  update_env("control/", controller.robots());
+  if(publish_real) { update_env("real/", controller.robots()); }
+
+  if(controller.robots().size() != mc_rtc::ROSBridge::nb_robot_publisher() / (publish_real ? 2 : 1))
```

Comment:
Won't this fail if we remove a robot and add a different robot? The size is the same thus we would not remove the extra robots. I think we should explicitly check whether the topic is still required.
Note that  in PR #462 we introduced a signal mechanism to notify changes to the robots that would simplify this kind of bookkeeping. We can revisit it once it's merged.